File: backend/src/modules/categories/domain/repositories/category.repository.ts
import { Category } from '../entities/category.entity';

export abstract class CategoryRepository {
  /** System categories (user_id IS NULL) plus this user's own, not soft-deleted. */
  abstract findAllForUser(userId: string): Promise<Category[]>;

  abstract findById(id: string): Promise<Category | null>;

  abstract findByUserAndName(
    userId: string,
    name: string,
  ): Promise<Category | null>;

  abstract create(data: {
    userId: string;
    name: string;
    icon: string;
    parentId: string | null;
  }): Promise<Category>;

  abstract update(
    id: string,
    changes: { name?: string; icon?: string; parentId?: string | null },
  ): Promise<Category>;

  abstract softDelete(id: string): Promise<void>;
}

File: mobile/lib/features/categories/domain/repositories/category_repository.dart
import '../entities/category.dart';

abstract class CategoryRepository {
  /// GET /categories — system categories plus the caller's own.
  Future<List<Category>> fetchAll();

  /// POST /categories — creates a custom category owned by the caller.
  Future<Category> create({
    required String name,
    required String icon,
    String? parentId,
  });
}
